modules/environment/login/default.nix

```diff
@@ -0,0 +1,94 @@
+# Licensed under GNU Lesser General Public License v3 or later, see COPYING.
+# Copyright (c) 2019 Alexander Sosedkin and other contributors, see AUTHORS.
+
+{ config, lib, pkgs, customPkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.environment.files;
+
+  login = pkgs.callPackage ./login.nix { inherit config; };
+
+  loginInner = pkgs.callPackage ./login-inner.nix { inherit config customPkgs; };
+in
+
+{
+
+  ###### interface
+
+  options = {
+
+    environment.files = {
+      login = mkOption {
+        type = types.package;
+        readOnly = true;
+        internal = true;
+        description = "Login script.";
+      };
+
+      loginInner = mkOption {
+        type = types.package;
+        readOnly = true;
+        internal = true;
+        description = "Login-inner script.";
+      };
+
+      prootStatic = mkOption {
+        type = types.package;
+        readOnly = true;
+        internal = true;
+        description = "proot-static package.";
+      };
+    };
+
+  };
+
+
+  ###### implementation
+
+  config = {
+
+    build.activation = {
+      installLogin = ''
+        if ! diff /bin/login ${login} > /dev/null; then
+          $DRY_RUN_CMD mkdir $VERBOSE_ARG --parents /bin
+          $DRY_RUN_CMD cp $VERBOSE_ARG ${login} /bin/.login.tmp
+          $DRY_RUN_CMD chmod $VERBOSE_ARG u+w /bin/.login.tmp
+          $DRY_RUN_CMD mv $VERBOSE_ARG /bin/.login.tmp /bin/login
+        fi
+      '';
+
+      installLoginInner = ''
+        if (test -e /usr/lib/.login-inner.new && ! diff /usr/lib/.login-inner.new ${loginInner} > /dev/null) || \
+            (! test -e /usr/lib/.login-inner.new && ! diff /usr/lib/login-inner ${loginInner} > /dev/null); then
+          $DRY_RUN_CMD mkdir $VERBOSE_ARG --parents /usr/lib
+          $DRY_RUN_CMD cp $VERBOSE_ARG ${loginInner} /usr/lib/.login-inner.tmp
+          $DRY_RUN_CMD chmod $VERBOSE_ARG u+w /usr/lib/.login-inner.tmp
+          $DRY_RUN_CMD mv $VERBOSE_ARG /usr/lib/.login-inner.tmp /usr/lib/.login-inner.new
+        fi
+      '';
+
+      installProotStatic = ''
+        if (test -e /bin/.proot-static.new && ! diff /bin/.proot-static.new ${cfg.prootStatic}/bin/proot-static > /dev/null) || \
+            (! test -e /bin/.proot-static.new && ! diff /bin/proot-static ${cfg.prootStatic}/bin/proot-static > /dev/null); then
+          $DRY_RUN_CMD mkdir $VERBOSE_ARG --parents /bin
+          $DRY_RUN_CMD cp $VERBOSE_ARG ${cfg.prootStatic}/bin/proot-static /bin/.proot-static.tmp
+          $DRY_RUN_CMD chmod $VERBOSE_ARG u+w /bin/.proot-static.tmp
+          $DRY_RUN_CMD mv $VERBOSE_ARG /bin/.proot-static.tmp /bin/.proot-static.new
+        fi
+      '';
+    };
+
+    environment.files = {
+      inherit login loginInner;
+
+      prootStatic =
```

Comment:
`path '/nix/store/bfxybmnxc39fjid3jsxsdfxnnszyijh0-proot-termux-unstable-2019-09-05-aarch64-unknown-linux-android' does not exist and cannot be created`

Reply:
Yes, you need to build and push proot into cachix as it gets fetched via its store path.

Reply:
I'm in progress, but that's something awkward.

Reply:
Built and pushed.

The reason is that it should point at a cross-compiled version, and hardcoding it is the only way to have the correct hash on-device?

There should be some way around that, otherwise that's pretty limiting: complicates forking to no end, updates become undeliverable without cachix...

Reply:
Either that way or uploading proot after each nixpkgs update to some cdn. We need a way to get proot-static updated on the device..

Reply:
The problem is, that currently we only push the build packages to cachix.

We need a build chain with aarch64 and i686 machines that build and push proot (and possible patched packages via overlays) to cachix, then we can use a nix expression instead of a hardcoded store path.

Reply:
For the build chains for aarch64 (https://docs.travis-ci.com/user/multi-cpu-architectures/) and i686 (https://unix.stackexchange.com/questions/292779/how-to-build-32-bit-application-on-64-bit-system-using-nix-package-manager/292829), there could be ways to achieve this via travis as build tool (as long as the build times are reasonable, otherwise we need to another solution, maybe get support by hydra.nixos.org?).

Reply:
But I'm not sure, how this works together with the fact, that the destination arch is an android phone..

Reply:
(something I can answer quickly) I'm fairly sure that it has to be build with a bionic toolchain (for the reasons not entirely clear to me), and that means x86_64 buildhost.

Reply:
Can you explain, what you mean by bionic buildchain?

Reply:
The one that is used to compile libraries for Android apps. Clang and bionic bundled by Google and distributed under the name of NDK. If proot is not compiled with/against it, it mysteriously terminates: #6.

Reply:
Ah I see, and compiling it with bionic does have a different nix store path than compliling it natively, right?

Reply:
Correct.